File: contents/stacks_and_queues/code/typescript/queue.ts
interface IQueue<T> {
  /**
   * `dequeue` removes first element from the queue and returns the same
   */
  dequeue(): T;
  /**
   * `enqueue` adds element to last of the queue and returns the size
   */
  enqueue(data: T): number;
  /**
   * `size` return size or length of the queue
   */
  size(): number;
  /**
   * `front` returns first element of the queue
   */
  front(): T;
}

class Queue<T> implements IQueue<T> {
  private readonly list: Array<T> = [];

  public enqueue(data: T) {
    return this.list.push(data);
  }

  public dequeue() {
    return this.list.shift();
  }

  public size() {
    return this.list.length;
  }

  public front() {
    return this.list[0];
  }
}

function exampleQueue() {
  const numberQueue = new Queue<number>();

  numberQueue.enqueue(4);
  numberQueue.enqueue(5);
  numberQueue.enqueue(9);

  console.log(numberQueue.dequeue());
  console.log(numberQueue.size());
  console.log(numberQueue.front());
}

exampleQueue();

File: contents/stacks_and_queues/code/java/Queue.java
import java.util.List;
import java.util.ArrayList;

public class QueueTest {
    
    public static void main(String[] args) {
	IQueue<Integer> intQueue = new Queue<>();

	intQueue.enqueue(4);
	intQueue.enqueue(5);
	intQueue.enqueue(9);
	
	System.out.println(intQueue.dequeue());
	System.out.println(intQueue.size());
	System.out.println(intQueue.peek());
    }

}


interface IQueue<T> {

   /*
    * 'dequeue' removes the first element from the queue and returns it
    */
    T dequeue();

   /*
    * 'enqueue' adds an element at the end of the queue and returns the new size
    */
    int enqueue(T element);
    

   /*
    * 'size' returns the size of the queue
    */
    int size();

   /*
    * 'peek' returns the first element of the queue without removing it
    */
    T peek();
}


public class Queue<T> implements  IQueue<T> {

    private List<T> list;

    public Queue() {
        this.list = new ArrayList();
    }

    public T dequeue() {
        return this.list.remove(0);
    }

    public int enqueue(T element) {
        this.list.add(element);
        return this.size();
    }

    public int size() {
        return this.list.size();
    }

    public T peek() {
        return this.list.get(0);
    }

}
